Source/santasyncservice/SNTPushClientNATS.mm

```diff
@@ -543,6 +582,174 @@ static void messageHandler(natsConnection *nc, natsSubscription *sub, natsMsg *m
+    [self sendCommandError:@"Command message has no data" toReplyTopic:replyTopic];
+    natsMsg_Destroy(msg);
+    return;
+  }
+
+  // Deserialize the message to SantaCommandRequest
+  // Note: We must extract all data from msg before destroying it, as NATS owns the message
+  // and will free it after this callback returns
+  santa::commands::v1::SantaCommandRequest command;
+  if (!command.ParseFromArray(data, dataLen)) {
+    LOGE(@"NATS: Failed to parse SantaCommandRequest from message on %@", msgSubject);
+    // Try to send error response, but don't fail if that also fails
+    [self sendCommandError:@"Failed to parse command" toReplyTopic:replyTopic];
+    natsMsg_Destroy(msg);
+    return;
+  }
+
+  // Destroy the message now - NATS owns it and will free it after callback returns anyway
+  // We've extracted all needed data (command, replyTopic, msgSubject) which are safe to capture
+  natsMsg_Destroy(msg);
+  msg = NULL;  // Prevent accidental use
+
+  // Process on message queue to serialize handling of messages
+  // Failures are logged but don't crash the client
+  dispatch_async(self.messageQueue, ^{
+    if (self.isShuttingDown) {
+      return;
+    }
+
+    // Dispatch to appropriate command handler based on oneof case
+    santa::commands::v1::SantaCommandResponse response;
+
+    switch (command.command_case()) {
+      case santa::commands::v1::SantaCommandRequest::kPing:
+        LOGI(@"NATS: Received PingRequest on %@", msgSubject);
+        response = [self handlePingCommand:command.ping()];
+        break;
+
+      case santa::commands::v1::SantaCommandRequest::COMMAND_NOT_SET:
+      default:
+        LOGE(@"NATS: Unknown command type on %@", msgSubject);
+        response.set_code(santa::commands::v1::SANTA_COMMAND_RESPONSE_CODE_ERROR);
+        break;
+    }
+
+    // Serialize and publish the response
+    // Serialization failures are logged but don't crash the client
+    std::string responseData;
+    if (!response.SerializeToString(&responseData)) {
+      LOGE(@"NATS: Failed to serialize command response (non-fatal)");
+      return;
+    }
+
+    // Publish response asynchronously - failures are logged but don't crash
+    dispatch_async(self.connectionQueue, ^{
+      if (![self isConnectionAlive]) {
+        LOGW(@"NATS: Cannot send command response - not connected (non-fatal)");
+        return;
+      }
+
+      natsStatus status =
+          natsConnection_Publish(self.conn, [replyTopic UTF8String], responseData.data(),
+                                 static_cast<int>(responseData.length()));
+      if (status != NATS_OK) {
+        LOGE(@"NATS: Failed to publish command response to %@: %s (non-fatal)", replyTopic,
+             natsStatus_GetText(status));
+      } else {
+        LOGD(@"NATS: Sent command response to %@ (code: %d)", replyTopic, response.code());
+      }
+    });
+  });
+}
+
+- (void)publishCommandResponse:(santa::commands::v1::SantaCommandResponseCode)resultCode
```

Comment:
I'm a little confused by this method, it only appears to be called by `sendCommandSuccess:toReplyTopic:` and `sendCommandError:toReplyTopic:`. The former is only called from the test, and the latter is called in a few places with error message strings but those aren't used?

In particular, the part that sets the ping response if the result code is SUCCESSFUL is unexpected given that the ping command handler already does this.